tahoe_idp/backend.py

```diff
@@ -127,3 +136,74 @@ def get_user_details(self, response):
         jwt_payload = self._get_payload(response)
         auth0_user = self.client.get_user(jwt_payload["email"])
         return self._build_user_details(jwt_payload=jwt_payload, auth0_user=auth0_user)
+
+
+class StudioTahoeIdpOAuth2(TahoeIdpOAuth2):
+    """
+    Backend for studio
+    """
+    def __init__(self, strategy=None, redirect_uri=None):
+        """
+        Special initialization to inject site_uuid in redirect_uri
+        """
+        super(StudioTahoeIdpOAuth2, self).__init__(strategy=strategy, redirect_uri=redirect_uri)
+        self.site = None
+        self.site_configuration = None
+
+        current_request = crum.get_current_request()
+        site_uuid = current_request.GET.get('site_uuid')
+
+        if site_uuid:
+            self.site = get_site_by_uuid(site_uuid=site_uuid)
+            self.site_configuration = getattr(self.site, "configuration", None)
+
+            # Inject site_uuid in redirect_uri
+            if self.redirect_uri and self.redirect_uri.endswith('/'):
+                self.redirect_uri = self.redirect_uri[:-1]
+            if self.redirect_uri:
+                self.redirect_uri = replace_query_param(self.redirect_uri, 'site_uuid', site_uuid)
+
+        if self.site_configuration:
+            # Tahoe: This might be the worst place to put the initialization logic, but it works for
+            #        Open edX so we're keeping it here to reduce performance impact
+            #        of `is_enabled_for_site()`
+            self.site_configuration.init_api_client_adapter(self.site)
+
+    @property
+    def client(self):
+        return Auth0ApiClient(site_configuration=self.site_configuration)
+
+    @property
+    def idp_domain(self):
+        return get_idp_domain(site_configuration=self.site_configuration)
+
+    @handle_http_errors
+    def do_auth(self, access_token, *args, **kwargs):
+        """
+        Override to allow only staff and admin users
+        """
+        result = super(StudioTahoeIdpOAuth2, self).do_auth(access_token=access_token, *args, **kwargs)
+
+        def redirect_to_lms_login(next='%2F'):
```

Comment:
Pretty minor, but it seems like this method would be a little more straightforward if `redirect_to_lms_login()` is just pulled out to be a method on the class rather than an inline function. If it ever needs to get more complicated, that would make it easier to test in isolation.